[crashpad] fix linux build

jarveson opened the pull request:
Crashpad creates a library named `libutil.so`, but this causes issues if it installs before python3, which causes py's configure script to use the incorrect and unrelated libutil.so when trying to check 'openpty' support. I've added a couple script replacements for the gn scripts which renames util -> crashpad_util. I also ran into a gcc linker error which looks to be fixed with a simple reordering ( linux build support still requires clang for crashpad lib build )

I'm not sure exactly what would be wanted or best solution here, so ive only renamed the one problematic library the port creates

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

jimwang118 commented:
I have the following problems when installing under Ubuntu22.10:
```
-- Building (x64-linux-dbg)...
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /usr/bin/ninja -C /mnt/crashpad/buildtrees/crashpad/x64-linux-dbg client client:common util:crashpad_util third_party/mini_chromium/mini_chromium/base handler:crashpad_handler
    Working Directory: /mnt/crashpad/buildtrees/crashpad/src/4c3e2d1c10-27b97453fa.clean
    See logs for more information:
      /mnt/crashpad/buildtrees/crashpad/build-x64-linux-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_ninja.cmake:3 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_build_ninja.cmake:24 (z_vcpkg_build_ninja_build)
  installed/x64-linux/share/vcpkg-gn/vcpkg_gn_install.cmake:94 (vcpkg_build_ninja)
  ports/crashpad/portfile.cmake:121 (vcpkg_gn_install)
  scripts/ports.cmake:147 (include)
```
[build-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/11631436/build-x64-linux-dbg-out.log)
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

jarveson commented:
> /bin/sh: 1: clang++: not found

yeap, so thats expected, you need clang installed on linux for the port to compile. This pr doesnt change that requirement at all

jimwang118 commented:
> > /bin/sh: 1: clang++: not found
> 
> yeap, so thats expected, you need clang installed on linux for the port to compile. This pr doesnt change that requirement at all

After I installed clang and compiled it, now the error message is as follows:
```
[1/5] CXX obj/util/net/net.http_transport_libcurl.o
FAILED: obj/util/net/net.http_transport_libcurl.o 
clang++ -MMD -MF obj/util/net/net.http_transport_libcurl.o.d -D_FILE_OFFSET_BITS=64 -DCRASHPAD_ZLIB_SOURCE_EXTERNAL -I../src/4c3e2d1c10-27b97453fa.clean -Igen -I../src/4c3e2d1c10-27b97453fa.clean/third_party/mini_chromium/mini_chromium -I../src/4c3e2d1c10-27b97453fa.clean/compat/linux -I../src/4c3e2d1c10-27b97453fa.clean/compat/non_win -I/mnt/crashpad/installed/x64-linux/include  -Wall -Wendif-labels -Werror -Wextra -Wextra-semi -Wheader-hygiene -Wnewline-eof -Wno-missing-field-initializers -Wno-unused-parameter -Wsign-compare -Wstring-conversion -Wvla -fno-exceptions -fno-rtti -fno-strict-aliasing -fobjc-call-cxx-cdtors -fstack-protector-all -fvisibility-inlines-hidden -fvisibility=hidden -fPIC -pthread -m64 -g -Wexit-time-destructors -Wimplicit-fallthrough -std=c++17 -c ../src/4c3e2d1c10-27b97453fa.clean/util/net/http_transport_libcurl.cc -o obj/util/net/net.http_transport_libcurl.o
../src/4c3e2d1c10-27b97453fa.clean/util/net/http_transport_libcurl.cc:17:10: fatal error: 'curl/curl.h' file not found
#include <curl/curl.h>
         ^~~~~~~~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
```

jarveson committed:
[crashpad] fix linux build
 - util lib renamed to avoid conflict with libutil
 - reorder library link order for gcc
 - add missing curl dependency on linux

Signed-off-by: jarves <jarveson@gmail.com>

jarveson commented:
Oh shoot, thanks, i guess the package has been dependent on an implicit curl dependency ( which my local project had )

Glancing at the crashpad docs <https://github.com/chromium/crashpad/blob/402d43192adc90b5bef3dabddb62655c7346533e/doc/developing.md#prerequisites> looks like only linux needs curl, so ive added that needed dependency for linux in vcpkg.json

I was able to recreate the curl issue on a more clean ubuntu install and after the change it compiled

jimwang118 commented:
Usage test pass with following triplets:

```
x64-linux
```

dan-shaw commented:
Is it possible to fix `python3` so that it doesn't automatically use `libutil`? In general, we don't want to rename upstream targets. Or is there another library that also creates `libutil`?

jarveson commented:
> Is it possible to fix python3 so that it doesn't automatically use libutil? In general, we don't want to rename upstream targets.

There isnt a way to fix it from pythons end afaik, (or at least anything trivial). Basically with one of the crashpad output libraries named 'libutil.so' it effectively hides the glibc 'libutil.so' (crashpads libutil is unrelated, just similar named) so the issue would also pop up with any other potential linux package that does a similar `-lutil` for openpty calls, or the other calls in that library. 

I understand if you don't want to rename upstream targets which is why i only touched that one (although tbh all the library outputs from crashpad are pretty poorly named to live in a shared output directory). The other option i had thought of would be trying to 'subdirectory' the library outputs, although i don't really see a way to with `vcpkg_gn_install` or with other edits to portfile

If theres a nicer way to subdirectory these im missing let me know. Otherwise this seemed at least the least 'lesser of evils' from reading packaging guidelines

jarveson commented:
Not sure why it would be remarked as draft with no response, but its clear this isnt a problem or issue that is worth fixing then
